r/src/table.cpp

```diff
@@ -327,6 +342,8 @@ std::shared_ptr<arrow::Table> Table__from_dots(SEXP lst, SEXP schema_sxp) {
   StopIfNotOk(
       arrow::r::CollectTableColumns(lst, schema, num_fields, infer_schema, columns));
 
+  StopIfNotOk(arrow::r::check_consistent_column_length(columns));
```

Comment:
I believe this will conflict with https://github.com/apache/arrow/pull/9615/ but I believe the resolution is trivial: just insert this line as the second to last line in the new definition of `Table__from_dots`